frontend/test/pytest/test_conditionals.py

```diff
@@ -30,8 +30,10 @@ def test_basic_cond_to_jaxpr(self):
         expected = """{ lambda ; a:i64[]. let
     b:bool[] = eq a 5
     c:i64[] = cond[
-      branch_jaxprs=[{ lambda ; a:i64[] b:i64[]. let c:i64[] = integer_pow[y=2] a in (c,) },
-                     { lambda ; a:i64[] b:i64[]. let c:i64[] = integer_pow[y=3] b in (c,) }]
+      branch_jaxprs=[{ lambda ; a:i64[] b_:i64[]. let c:i64[] = integer_pow[y=2] a in (c,) },
```

Comment:
This is our constant-padding code:) See [here](https://github.com/PennyLaneAI/catalyst/blob/f03253d8e5ffeee5b91ba79514057d8d813bc874/frontend/catalyst/utils/jax_extras.py#L269). We modify Jaxpr programs representing the cond branches using a simple copy-all-together algorithm. We make them to have the similar sets of constants which seems to be one of the StableHLO requirements.

Reply:
Thanks! 
> We make them to have the similar sets of constants which seems to be one of the StableHLO requirements.

What changed that made this padding necessary? My thinking is that if this was a requirement of StableHLO, it would have already been a requirement before the dynamic API no?

The docstring of the linked function is also a good example of code that could use improved documentation:
> """Align the constants of Jaxpr programs. Return the list of corresponding programs accepting
>      same constants."""

In this case, what does it mean to "align the constants of a Jaxpr program"? And almost more importantly, _why_ is this being done?

Reply:
> if this was a requirement of StableHLO, it would have already been a requirement before the dynamic API no?

Yes, it was. We used a different function for this, which didn't support dynamic api (but it did support some other modes that we don't need). I implemented a version which does support dynamism.

>  what does it mean to "align the constants of a Jaxpr program"?  And almost more importantly, why is this being done?

Basically, if we have two programs with constants  `a` and `b`, we transform them into the programs `a, b_` and `a_, b` where constants with `_` suffix are unused. Yes, it was the requirement of StableHLO before this PR.

Reply:
Could you elaborate on how these two programs are used? Are you saying where we used to have a single jaxpr, there are now two?

Reply:
No, by programs I meant the programs encoding the conditional branches. Users may write arbitrary many branches, each with its own set of constants (and, potentially, inputs, but for conditional we have zero inputs - it does make our life easier). We need to unify the signatures: they need to be the same so our StableHLO lowering can work. We do this unification by adding all constants to all programs. This is what I tried to say in the previous post. In the end we have the same number of programs, equal to the number of conditional branches, with equal set of constants.

Reply:
Got it, I think I get the picture now. The conditional requires that the inputs for each branch are the same, which we have always performed by using existing JAX functionality. Now you wrote your own function to perform the equivalent task, but with dynamic array support. In your own implementation, you marked the unused constants for a program with an underscore, which is the change visible in this diff? :)

Reply:
Just not inputs (cond branches have no inputs!), but constants. Yes, I wrote my own implementation of the padding algorithm which takes the variable shapes of _constant outer tracers_ (mind the discussion about constants!) into account.